python/ray/data/datasource/file_based_datasource.py

```diff
@@ -202,11 +201,17 @@ def _resolve_paths_and_filesystem(
         raise ValueError("Must provide at least one path.")
 
     if filesystem and not isinstance(filesystem, FileSystem):
+        err_msg = f"The filesystem passed must either conform to " \
+                  f"pyarrow.fs.FileSystem, or " \
+                  f"fsspec.spec.AbstractFileSystem. The provided " \
+                  f"filesystem was: {filesystem}"
+        try:
+            import fsspec
+        except ModuleNotFoundError:
+            raise TypeError(err_msg)
```

Comment:
This is not the right error to raise!

Reply:
What error do you think should be raised? This means they passed in a fs object that isn't a pa.fs or fsspec fs.

Reply:
@ericl Agreed with @wuisawesome here! They only reach this import error if the passed filesystem is not a pyarrow filesystem, and given that the only other thing we support are fsspec filesystems, being unable to import `fsspec` suggests implies that the provided `filesystem` can't be an fsspec filesystem, so the `filesystem` arg is the wrong type, so we should raise a type error here.

If this might be a point of confusion for future devs reading this logic, maybe we should have a `_is_fsspec_fs()` helper:
```python
if not _is_fsspec_fs(filesystem):
    err_msg = f"The filesystem passed must either conform to " \
              f"pyarrow.fs.FileSystem, or " \
              f"fsspec.spec.AbstractFileSystem. The provided " \
              f"filesystem was: {filesystem}"
    raise TypeError(err_msg)

filesystem = PyFileSystem(FSSpecHandler(filesystem))

# ...


def _is_fsspec_fs(fs: Any):
    try:
        import fsspec
    except ModuleNotFoundError:
        return False

    return isinstance(filesystem, fsspec.spec.AbstractFileSystem)
```

Not sure if it's worth it though, could probably just add a comment above this line about why we're raising a `TypeError` here.

Reply:
```suggestion
            # If filesystem is not a pyarrow filesystem and fsspec isn't
            # installed, then filesystem is neither a pyarrow filesystem nor
            # an fsspec filesystem, so we raise a TypeError.
            raise TypeError(err_msg)
```